### PR TITLE
#30254: Adds a "minimum_version" setting to bundle framework requirements.

### PR DESCRIPTION
**Pull request by thebeeland**
If a bundle requires a minimum version of a specific framework, that can be specified in its `info.yml`.

```
# the frameworks required to run this app
frameworks:
    - {"name": "tk-framework-widget", "version": "v0.2.x", "minimum_version": "v0.2.3"}
```

When `tank updates` is run and a bundle is updated, an additional step is taken to ensure that any required frameworks are also updated to at least the `minimum_version` requirement specified in the bundle's `info.yml`. This update occurs automatically, and without user confirmation, to ensure that the newly-updated bundle will function properly.

### CODE REVIEW

File: python/tank/commands/update.py [hunk 1]
```diff
@@ -615,3 +624,66 @@ def _check_item_update_status(environment_obj, engine_name=None, app_name=None,
 
     return data
 
+
+def _get_framework_requirements(environment, descriptor):
+    """
+    Gets a list of framework updates that will be required for the given
```
Comment (josh-t) [inline python/tank/commands/update.py]: s/Gets/Returns/

File: python/tank/platform/validation.py [hunk 2]
```diff
@@ -156,18 +164,50 @@ def validate_and_return_frameworks(descriptor, environment):
...
+                # out to v2.1.0 and we have a minimum required version of v2.1.5,
+                # then we haven't actually found a compatible framework.
+                #
+                # If the descriptor for the framework doesn't contain a concrete
+                # version number (example: a "dev" descriptor won't have one), we
+                # simply skip the minimum-required version check. There's nothing
+                # we can do to confirm, so it's best to trust the config. It's
+                # likely that anyone using a non-app_store descriptor understands
+                # the caveats that come with them.
+                fw_version = fw_desc.version
```
Comment (josh-t) [inline python/tank/platform/validation.py]: Should we add a log statement here if the `fw_version` is None? To point out the scenario you describe above?

File: python/tank/platform/validation.py [hunk 3]
```diff
@@ -156,18 +164,50 @@ def validate_and_return_frameworks(descriptor, environment):
...
+
+                if min_version and fw_version:
+                    # Check to make sure the version strings aren't malformed.
+                    # We expect v\d+.\d+.\d+ pattern. Group 1 of the match will
+                    # be the version without the "v" at the head.
+                    req_match = re.match(fw_version_pattern, min_version)
```
Comment (josh-t) [inline python/tank/platform/validation.py]: Same comment from above. I feel like this kind of stuff should be in a util somewhere...

Aha! Would this work? 

https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/util/version.py#L24

File: python/tank/platform/validation.py [hunk 4]
```diff
@@ -156,18 +164,50 @@ def validate_and_return_frameworks(descriptor, environment):
...
+                    fw_match = re.match(fw_version_pattern, fw_version)
+
+                    # If either were malformed, then we just have to skip the check.
```
Comment (josh-t) [inline python/tank/platform/validation.py]: log warning here?

File: python/tank/platform/validation.py (hunk 3 above)
Comment (thebeeland) [inline python/tank/platform/validation.py]: Aha!

File: python/tank/platform/validation.py (hunk 2 above)
Comment (thebeeland) [inline python/tank/platform/validation.py]: Sticking with just a warning in the case where the comparison can't be made for a technical reason.

**Comment by thebeeland** [comment]
I've addressed everything, I believe, plus added a test case for the new feature.

Below is output from `tank updates` that shows what it looks like with no framework update requirements, then WITH framework update requirements, plus the output when that is accepted.

```
----------------------------------------------------------------------
Engine tk-maya (Environment asset)
/----------------------------------------------------------------------
| Item:        tk-maya v0.7.8
|
| Description: Shotgun Integration in Maya
|
| Change Log:  Recreates Maya event handler and resets the Maya
|              project when changing context.
\----------------------------------------------------------------------
Update to the above version? [Yna?]n

App tk-multi-about (Engine tk-maya, Environment asset)
/----------------------------------------------------------------------
| Item:             tk-multi-about v0.2.0
|
| Description:      Shows a breakdown of your current environment
|                   and configuration.
|
| Change Log:       No details provided.
|
| Required Updates: tk-framework-widget_v0.2.x
\----------------------------------------------------------------------
Update to the above version? [Yna?]y


Framework tk-framework-widget_v0.2.x (Environment asset)
/----------------------------------------------------------------------
| Item:        tk-framework-widget v0.2.4
|
| Description: Collection of standard Shotgun Pipeline Toolkit
|              Related Q-Widgets.
|
| Change Log:  Updates the framework dependencies to the latest
|              major version.
\----------------------------------------------------------------------
```

**Comment by josh-t** [comment]
Output looks awesome!

**Pull request merged**